### PR TITLE
Pim 3807 behat user management

### PR DESCRIPTION
**Pull request by jmleroux**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | n
| BC breaks?           | n
| CI currently passes? | y
| Tests pass?          | y
| Scenarios pass?      | y
| Checkstyle issues?*  | n
| PMD issues?**        |
| Changelog updated?   | n
| Fixed tickets        | PIM-3807
| DB schema updated?   | n
| Migration script?    | n
| Doc PR               | n

### CODE REVIEW

File: features/Context/FixturesContext.php [hunk 1]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+     */
+    public function checkUserAssociationExists($username, $searchedLabel = null, $associationType = null)
+    {
+        /** @var User $userEntity */
+        $userEntity = $this->getEntityOrException('User', ['username' => $username]);
+        if ($searchedLabel && $associationType == 'group' && !$userEntity->hasGroup($searchedLabel)) {
+            throw new \InvalidArgumentException("The user $username does not belong to the '$searchedLabel' group");
+        }
+        if ($searchedLabel && $associationType == 'role' && !$userEntity->hasRole($searchedLabel)) {
+            var_dump($searchedLabel);
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: var_dump() ;)

File: features/Context/FixturesContext.php [hunk 2]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+     *
+     * @Then /^the user "([^"]+)" should be in the "(?P<groupname>[^"]+)" (group)$/
+     * @Then /^the user "([^"]+)" should have the "([^"]+)" (role)$/
+     */
+    public function checkUserAssociationExists($username, $searchedLabel = null, $associationType = null)
+    {
+        /** @var User $userEntity */
+        $userEntity = $this->getEntityOrException('User', ['username' => $username]);
+        if ($searchedLabel && $associationType == 'group' && !$userEntity->hasGroup($searchedLabel)) {
+            throw new \InvalidArgumentException("The user $username does not belong to the '$searchedLabel' group");
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: Do not insert directly variables inside your string. Prefer sprintf()

File: features/Context/FixturesContext.php [hunk 3]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+        }
+        if ($searchedLabel && $associationType == 'role' && !$userEntity->hasRole($searchedLabel)) {
+            var_dump($searchedLabel);
+            throw new \InvalidArgumentException("The user $username does not have the '$searchedLabel' role");
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: Idem, use sprintf()

File: features/Context/FixturesContext.php [hunk 4]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+        }
+        return $userEntity;
+    }
+
+    /**
+     * @param string $username
+     * @param string $groupname
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: PHPDoc is not valid

File: features/Context/FixturesContext.php [hunk 5]
```diff
@@ -1374,6 +1391,64 @@ public function getUser($username)
...
+     *
+     * @return User
+     *
+     * @Then /^the user "([^"]+)" should not be in the "(?P<groupname>[^"]+)" (group)$/
+     * @Then /^the user "([^"]+)" should not have the "([^"]+)" (role)$/
+     */
+    public function checkUserAssociationDoNotExist($username, $searchedLabel = null, $associationType = null)
+    {
+        return $this->checkUserAssociationExists($username, $searchedLabel, $associationType) ? false : true;
```
Comment (grena) [inline features/Context/FixturesContext.php]: `return false === $this->checkUserAssociationExists($username, $searchedLabel, $associationType)` seems clearer than a ternary IMO

File: features/Context/FixturesContext.php [hunk 6]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+    /**
+     * @param string $username
+     * @param string $groupname
+     *
+     * @return User
+     *
+     * @Then /^the user "([^"]+)" should not be in the "(?P<groupname>[^"]+)" (group)$/
+     * @Then /^the user "([^"]+)" should not have the "([^"]+)" (role)$/
+     */
+    public function checkUserAssociationDoNotExist($username, $searchedLabel = null, $associationType = null)
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: There are differences between variables and regexp

File: features/Context/FixturesContext.php [hunk 7]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+    {
+        return $this->checkUserAssociationExists($username, $searchedLabel, $associationType) ? false : true;
+    }
+
+    /**
+     * @param string $username
+     * @param string $groupname
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: PHPDoc is not valid

File: features/Context/FixturesContext.php [hunk 8]
```diff
@@ -1374,6 +1391,65 @@ public function getUser($username)
...
+     *
+     * @return bool
+     *
+     * @Then /^the user "([^"]+)" should be in (\d+) (group)s?$/
+     * @Then /^the user "([^"]+)" should(?: still)? have (\d+) (role)s?$/
+     */
+    public function checkUserAssociationsCount($username, $count, $associationType)
```
Comment (skeleton) [inline features/Context/FixturesContext.php]: There are differences between variables and regexp

File: features/Context/NavigationContext.php [hunk 9]
```diff
@@ -203,12 +213,40 @@ public function iAmOnTheEntityEditPage($identifier, $page)
...
+     *
+     * @Given /^I edit the "([^"]+)" user role$/
+     */
+    public function iEditTheUserRole($label)
+    {
+        $page = 'UserRole';
+        $getter = sprintf('get%s', $page);
+        /** @var VersionableInterface $entity */
+        $entity = $this->getFixturesContext()->$getter($label);
+        $this->openPage(sprintf('UserRole edit', $page), array('id' => $entity->getId()));
```
Comment (grena) [inline features/Context/NavigationContext.php]: please use `[]` for your arrays :)

File: features/Context/FixturesContext.php (hunk 6 above)
Comment (jmleroux) [inline features/Context/FixturesContext.php]: I will remove tne named capturing group as Behat 2 does not take it into account

File: features/Context/Page/AssociationType/Creation.php [hunk 10]
```diff
@@ -13,4 +13,8 @@
  */
 class Creation extends Form
 {
+    /**
+     * @var string
+     */
+    protected $path = '/configuration/association-type/create';
```
Comment (grena) [inline features/Context/Page/AssociationType/Creation.php]: inline PHPdoc

File: features/Context/Page/UserGroup/Creation.php [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * @author    Jean-Marie Leroux <jean-marie.leroux@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Creation extends Form
+{
+    /**
+     * @var string $path
+     */
+    protected $path = '/user/group/create';
```
Comment (grena) [inline features/Context/Page/UserGroup/Creation.php]: We use inline syntax for simple attribute PHPdoc

File: features/user/create_and_delete_group.feature [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    And I confirm the deletion
+    When I click on the "Delete" action of the row which contains "DummyGroup3"
+    Then I should see "Delete Confirmation"
+    And I confirm the deletion
+
+    Given I am on the user groups page
+    Then there should be a "Redactor" user group
+    When I click on the "Delete" action of the row which contains "Redactor"
+    Then I should see a confirm dialog with the following content:
+      | title   | Delete Confirmation                             |
```
Comment (skeleton) [inline features/user/create_and_delete_group.feature]: align

File: features/user/create_and_delete_role.feature [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    Given I am on the user roles creation page
+    When I save the role
+    Then I should see validation tooltip "This value should not be blank."
+
+  @javascript
+  Scenario: Successfully delete a role after confirmation
+    Given I am on the user roles page
+    When I click on the "Delete" action of the row which contains "Dummy role"
+    Then I should see a confirm dialog with the following content:
+      | title   | Delete Confirmation                             |
```
Comment (skeleton) [inline features/user/create_and_delete_role.feature]: align

File: features/user/create_and_delete_role.feature [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+      | content | Are you sure you want to delete this role? |
+    When I confirm the deletion
+    Then I should not see "Dummy role"
+
+    When I click on the "Delete" action of the row which contains "Administrator"
+    Then I should see a confirm dialog with the following content:
+      | title   | Delete Confirmation                             |
```
Comment (skeleton) [inline features/user/create_and_delete_role.feature]: align

File: features/user/create_and_delete_group.feature [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    And I should see "DummyGroup1"
+    And I should see "DummyGroup3"
+    And I should not see "DummyGroup2"
+
+    Given I am on the user groups creation page
+    When I save the group
+    Then I should see "This value should not be blank."
+
+  @javascript
+  Scenario: delete created groups for database integrity
```
Comment (grena) [inline features/user/create_and_delete_group.feature]: Missing first letter uppercase

File: features/Context/FixturesContext.php (hunk 8 above)
Comment (filipsalpe) [inline features/Context/FixturesContext.php]: (?: still)? is a non-matching group, looks good to me

File: features/Context/NavigationContext.php [hunk 16]
```diff
@@ -610,7 +663,7 @@ public function openPage($page, array $options = array())
     }
 
     /**
-     * @return Page
+     * @return Form
```
Comment (filipsalpe) [inline features/Context/NavigationContext.php]: Page seems to be the correct one here

File: features/Context/FixturesContext.php [hunk 17]
```diff
@@ -1374,6 +1390,67 @@ public function getUser($username)
...
+     * @param string $associationType Can be 'group' or 'role'
+     *
+     * @return bool
+     *
+     * @Then /^the user "([^"]+)" should be in (\d+) (group)s?$/
+     * @Then /^the user "([^"]+)" should(?: still)? have (\d+) (role)s?$/
+     */
+    public function checkUserAssociationsCount($username, $count, $associationType)
+    {
+        /** @var User $userEntity */
```
Comment (filipsalpe) [inline features/Context/FixturesContext.php]: :-1:

File: features/user/create_and_delete_group.feature [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+  @javascript
+  Scenario: Successfully create a user group
+    Given I am on the user groups creation page
+    And I fill in the following information:
+      | Name     | DummyGroup1 |
+      | Owner    | Main        |
+    When I save the group
+    Then there should be a "DummyGroup1" user group
+
```
Comment (filipsalpe) [inline features/user/create_and_delete_group.feature]: No need for blank lines within a scenario

File: features/Context/FixturesContext.php [hunk 19]
```diff
@@ -1374,6 +1390,67 @@ public function getUser($username)
...
+        $userEntity = $this->getEntityOrException('User', ['username' => $username]);
+        $check = false;
+        if ($associationType == 'group' && count($userEntity->getGroupNames()) == $count) {
+            $check = true;
+        } elseif ($associationType == 'role' && count($userEntity->getRoles()) == $count) {
+            $check = true;
+        }
+        return $check;
```
Comment (filipsalpe) [inline features/Context/FixturesContext.php]: Can this step fail somehow?

File: features/Context/Loader/UserLoader.php [hunk 20]
```diff
@@ -105,6 +104,7 @@ protected function createUser($username, array $data)
         // Following to fix a cascade persist issue on UserApi occuring only during Behat Execution
         $this->getUserManager()->getStorageManager()->clear('Oro\Bundle\UserBundle\Entity\User');
         $this->getUserManager()->getStorageManager()->clear('Oro\Bundle\UserBundle\Entity\UserApi');
+        $this->getUserManager()->getStorageManager()->clear('OroEmail\Cache\OroEmailBundle\Entity\EmailAddressProxy');
```
Comment (filipsalpe) [inline features/Context/Loader/UserLoader.php]: No need for this anymore

File: features/user/create_and_delete_group.feature [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    And I should see "DummyGroup3"
+    And I should not see "DummyGroup2"
+    Given I am on the user groups creation page
+    When I save the group
+    Then I should see "This value should not be blank."
+
+  @javascript
+  Scenario: Delete previously created groups
+    Given I am on the user groups page
+    When I click on the "Delete" action of the row which contains "DummyGroup1"
```
Comment (filipsalpe) [inline features/user/create_and_delete_group.feature]: This group should not exist at this point.

File: features/Context/FixturesContext.php [hunk 22]
```diff
@@ -1374,6 +1389,70 @@ public function getUser($username)
...
+     * @param string $associationType Can be 'group' or 'role'
+     *
+     * @return bool
+     *
+     * @Then /^the user "([^"]+)" should not be in the "([^"]+)" (group)$/
+     * @Then /^the user "([^"]+)" should not have the "([^"]+)" (role)$/
+     */
+    public function checkUserAssociationDoNotExist($username, $searchedLabel = null, $associationType = null)
+    {
+        return !$this->checkUserAssociationExists($username, $searchedLabel, $associationType);
```
Comment (filipsalpe) [inline features/Context/FixturesContext.php]: This won't work, you'll get an exception

File: features/Context/NavigationContext.php [hunk 23]
```diff
@@ -588,7 +626,6 @@ public function iShouldBeOnTheProductEditPage(Product $product)
     public function iRefreshCurrentPage()
     {
         $this->getMainContext()->reload();
-        $this->wait();
```
Comment (filipsalpe) [inline features/Context/NavigationContext.php]: This could have some undesired impacts..

File: features/Context/excluded_tables.yml [hunk 24]
```diff
@@ -4,7 +4,7 @@ excluded_tables:
     - acl_object_identities
     - acl_object_identity_ancestors
     - acl_security_identities
-    - oro_access_group
-    - oro_access_role
     - oro_config
     - oro_config_value
+    - oro_access_group
```
Comment (filipsalpe) [inline features/Context/excluded_tables.yml]: Alphabetic order was nicer :)

File: features/user/create_edit_and_delete_group.feature [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+    Then there should be a "DummyGroup3" user group
+    When I am on the user groups page
+    Then the grid should contain 5 elements
+    And I should see "DummyGroup1"
+    And I should see "DummyGroup3"
+    And I should not see "DummyGroup2"
+
+    Given I am on the user groups creation page
+    When I save the group
+    Then I should see "This value should not be blank."
```
Comment (filipsalpe) [inline features/user/create_edit_and_delete_group.feature]: As we don't reset the groups page, manual cleanup is necessary - you should drop these 2 groups in the scenario

File: features/user/create_edit_and_delete_group.feature [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+
+  @javascript
+  Scenario: Successfully edit a user group
+    Given I am on the user groups page
+    Then there should be a "DummyGroup3" user group
```
Comment (filipsalpe) [inline features/user/create_edit_and_delete_group.feature]: Shouldn't exist at this point

File: features/user/create_edit_and_delete_group.feature [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+    And I save and close
+    Then there should be a "DummyGroup4" user group
+    And I should see "DummyGroup4"
+    And I should not see "DummyGroup3"
+
+
+  @javascript
+  Scenario: Successfully delete groups
+    Given I am on the user groups page
+    When I click on the "Delete" action of the row which contains "DummyGroup1"
```
Comment (filipsalpe) [inline features/user/create_edit_and_delete_group.feature]: Same as above, should not exist

**Pull request merged**